cmd/system-probe/api/client/client.go

```diff
@@ -7,10 +7,111 @@
 package client
 
 import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
 	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/cmd/system-probe/config/types"
+	"github.com/DataDog/datadog-agent/pkg/util/funcs"
+)
+
+var (
+	// ErrNotImplemented is an error used when system-probe is attempted to be accessed on an unsupported OS
+	ErrNotImplemented = errors.New("system-probe unsupported")
 )
 
 // Get returns a http client configured to talk to the system-probe
-func Get(socketPath string) *http.Client {
-	return newSystemProbeClient(socketPath)
+var Get = funcs.MemoizeArgNoError[string, *http.Client](get)
+
+func get(socketPath string) *http.Client {
+	return &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			MaxIdleConns:          2,
+			IdleConnTimeout:       idleConnTimeout,
+			DialContext:           DialContextFunc(socketPath),
+			TLSHandshakeTimeout:   1 * time.Second,
+			ResponseHeaderTimeout: 5 * time.Second,
+			ExpectContinueTimeout: 50 * time.Millisecond,
+		},
+	}
+}
+
+// GetCheck returns data unmarshalled from JSON to T, from the specified module at the /<module>/check endpoint.
+func GetCheck[T any](client *http.Client, module types.ModuleName) (T, error) {
+	var data T
+	req, err := http.NewRequest("GET", ModuleURL(module, "/check"), nil)
+	if err != nil {
+		return data, err
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return data, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return data, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return data, fmt.Errorf("non-ok status code: url %s, status_code: %d, response: `%s`", req.URL, resp.StatusCode, string(body))
+	}
+
+	err = json.Unmarshal(body, &data)
+	return data, err
+}
+
+func constructURL(module string, endpoint string) string {
+	u, _ := url.Parse("http://sysprobe")
+	if module != "" {
+		u = u.JoinPath(module)
+	}
+	path, query, found := strings.Cut(endpoint, "?")
+	u = u.JoinPath(path)
+	if found {
+		u.RawQuery = query
+	}
+	return u.String()
+}
+
+// URL constructs a system-probe URL for a module-less endpoint.
+func URL(endpoint string) string {
+	return constructURL("", endpoint)
+}
+
+// DebugURL constructs a system-probe URL for the debug module and endpoint.
+func DebugURL(endpoint string) string {
+	return constructURL("debug", endpoint)
+}
+
+// ModuleURL constructs a system-probe ModuleURL given the specified module and endpoint.
+func ModuleURL(module types.ModuleName, endpoint string) string {
+	return constructURL(string(module), endpoint)
+}
+
+// ReadAllResponseBody reads the entire HTTP response body as a byte slice
+func ReadAllResponseBody(resp *http.Response) ([]byte, error) {
+	// if we are not able to determine the content length
+	// we read the whole body without pre-allocation
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Out of curiosity, what is the average expected `content-length` ? It is generally recommended to have a maximum allocation size to avoid allocating an enormous memory size if the received `content-length` is erroneous.

Reply:
It would depend on the use cases. I was preserving code which was already doing this. I think `io.LimitedReader` helps with this, but that should probably be left to a future PR.